src/main.rs

```diff
@@ -239,6 +247,44 @@ fn build_gga(opt: &Cli) -> Command {
     }
 }
 
+struct AreaIDParams {
+    a: f32,
+    b: f32,
+    offset: i32,
+}
+
+fn get_area_id_parameters(lat: f32) -> AreaIDParams {
+    if lat > 60.0 && lat <= 75.0 {
+        AreaIDParams {
+            a: 0.04,
+            b: 0.02,
+            offset: 0,
+        }
+    } else if lat > -60.0 && lat <= 60.0 {
+        AreaIDParams {
+            a: 0.02,
+            b: 0.02,
+            offset: -6750000,
+        }
+    } else if lat > -75.0 && lat <= -60.0 {
+        AreaIDParams {
+            a: 0.04,
+            b: 0.02,
+            offset: 54000000,
+        }
+    } else {
+        unimplemented!("Invalid latitude {lat}")
+    }
+}
+
+fn area_id(lat: f32, lon: f32) -> u32 {
+    let params = get_area_id_parameters(lat);
+
+    ((360.0 / params.a) * (75.0 - lat) / params.b) as u32
+        + ((lon + 180.0) / params.a) as u32
+        + params.offset as u32
```

Comment:
personally i prefer using the num_traits cast function for these sorts of things just for added safety. (plus on old versions of rust casting out of bound floats could cause UB.)

```suggestion
    use num_traits::cast;
    cast::<_,u32>((360.0 / params.a) * (75.0 - lat) / params.b).unwrap()
        + cast::<_,u32>((lon + 180.0) / params.a).unwrap()
        +  cast::<_,u32>(params.offset)
```

Reply:
Casts like this can definitely invoke platform specific behavior, but the algorithm as specified explicitly wants to use C style conversion from the floating point intermediates to integers which can have those platform specific behaviors so I think I'll keep this casting approach here (but thanks for showing me these num-trait casts!!).

This is actually from a few months ago, and I just looked at a more recent C implementation and they are using explicit truncations before casting from float to int. I'll try re-writing this with the truncations and see if it makes things a little more clear.

Reply:
My concern was more about f32 that might be negative or have a value larger than u32::max().

Reply:
Ah I see what you mean. The little calculation here _should_ produce valid positive integer values if the latitude/longitude are within the accepted ranges, and a little bit above this when picking out the parameters if the latitude/longitude are not within the accepted ranges I have an `unimplemented!()` call which should crash the entire program before we end up producing invalid values.

Reply:
Oh wait, it is only validating the latitude. Let me add a quick check on both values.